data/abraca.1
.\" Copyright (C) 2009  Daniel Svensson <dsvensson@gmail.com>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer
.\"    in this position and unchanged.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.TH ABRACA 1 "2009-08-17"
.SH NAME
abraca \- XMMS2 client for the GNOME desktop
.SH SYNOPSIS
.B abraca
.SH DESCRIPTION
Abraca is a GTK2 client for the XMMS2 music player. It is designed with
collections in mind, which makes managing your music a breeze.
.SH FILES
.TP
.I $XDG_CONFIG_HOME/xmms2/clients/abraca.conf
A configuration file containing basic configuration values for
.I abraca.
.SH ENVIRONMENTAL VARIABLES
.TP
.I XMMS_PATH
The URL path to
.B xmms2d
in the format unix:///path/to/socket, or tcp://a.b.c.d:port.
.TP
.I XDG_CONFIG_HOME
The path where the Abraca configuration file is located.

.SH SEE ALSO
xmms2d(8), http://xmms2.xmms.org, http://abraca.github.com/Abraca
.SH HISTORY
Abraca was first written in C by Sebastian Sareyko, but was later
rewritten in Vala by the current Abraca Team.
.SH AUTHORS
Abraca was written by the Abraca Team.
.PP
This manual page was written by Daniel Svensson
<dsvensson@gmail.com>


debian/changelog
abraca (0.7.0-1) unstable; urgency=low

  * Added DM-Upload-Allowed field
  * Bumped Standards-Version to 3.9.3 (use copyright-format 1.0)
  * Updated valac-0.10 B-D to valac-0.14 (Closes: #663292)
  * Fixed DM-Allow-Upload field in d/control
  * Imported Upstream version 0.7.0
  * Removed libgtk2.0-dev in B-D in favor of newer libgtk-3-dev
  * Bump libxmmsclient-glib-dev B-D to (>= 0.8)
  * Added 01-change-icon.diff patch for using different icon
  * Added myself to d/copyright and bumped year
  * Fixed d/watch
  * Changed Homepage field in d/control
  * Fixed missing-field-in-dep5-copyright lintian message
  * Added missing copyright files and licenses

 -- Fabrizio Regalli <fabreg@fabreg.it>  Tue, 20 Mar 2012 23:52:52 +0100

abraca (0.6.0-3.1) unstable; urgency=low

  * Non-Maintainer Upload.
  * debian/rules:
    - Build using the bundled version of scons.
      Fixes FTBFS on amd64 machines. (Closes: #642678)

 -- Andrea Veri <and@debian.org>  Sun, 25 Sep 2011 15:58:38 +0200

abraca (0.6.0-3) unstable; urgency=low

  * New maintainer (Closes: #619778).
  * Added upstream changelog.

 -- Fabrizio Regalli <fabreg@fabreg.it>  Tue, 05 Apr 2011 15:28:30 +0200

abraca (0.6.0-2) unstable; urgency=low

  * Upload to unstable.
  * Explicitly build depend on valac-0.10.

 -- Benjamin Drung <bdrung@debian.org>  Wed, 16 Feb 2011 20:18:01 +0100

abraca (0.6.0-1) experimental; urgency=low

  * New upstream release.
  * Update my email address.
  * Bump Standards-Version to 3.9.1 (no changes required).
  * Update watch file to detect gz and bz2.
  * Drop all patches (applied upstream).
  * Bump valac build dependency to >= 0.10.0.

 -- Benjamin Drung <bdrung@debian.org>  Sat, 22 Jan 2011 19:44:55 +0100

abraca (0.5.0-3) unstable; urgency=low

  * Bump libxmmsclient-glib-dev dependency to >= 0.7DrNo.

 -- Benjamin Drung <bdrung@ubuntu.com>  Wed, 14 Apr 2010 14:05:57 +0200

abraca (0.5.0-2) unstable; urgency=low

  * Fix FTBFS with vala 0.8.0.

 -- Benjamin Drung <bdrung@ubuntu.com>  Tue, 13 Apr 2010 23:46:29 +0200

abraca (0.5.0-1) unstable; urgency=low

  * New upstream release.
    - Compatible to vala 0.7.10 and xmms2 0.7DrNo (Closes: #571430).
    - Information dialog is closable (Closes: #565734).
    - Abraca launches xmms2d if unable to connect (Closes: #440678).
  * Update debian/copyright.
  * Add libgee-dev to Build-Depends.
  * Bump Standards-Version 3.8.4 (no changes required).

 -- Benjamin Drung <bdrung@ubuntu.com>  Sat, 03 Apr 2010 11:32:47 +0200

abraca (0.4.3-1) unstable; urgency=low

  * New upstream release.
    - Compatible to vala 0.7.9 (Closes: #564354).

 -- Benjamin Drung <bdrung@ubuntu.com>  Fri, 15 Jan 2010 10:07:35 +0100

abraca (0.4.2-1) unstable; urgency=low

  * New upstream release.
    - Compatible to vala 0.7.8 (Closes: #556812).
  * Drop all patches and the patch system.
  * Switch to source format 3.0 (quilt).
  * Bump Build-Depends version of valac to >= 0.7.8.
  * Update debian/watch file.

 -- Benjamin Drung <bdrung@ubuntu.com>  Tue, 01 Dec 2009 00:06:03 +0100

abraca (0.4.1-2) unstable; urgency=low

  * Add debian/patches/fix-packaging-mistake.patch to raise the package
    to the real 0.4.1 release.

 -- Benjamin Drung <bdrung@ubuntu.com>  Fri, 16 Oct 2009 00:38:40 +0200

abraca (0.4.1-1) unstable; urgency=low

  * New upstream release.
    - Compatible to vala 0.7.7 (Closes: #549787).
  * Drop fix-binaryblob-comment.patch (applied upstream).
  * Drop version-fix.patch (no longer required).
  * Remove patch system (no patches remaining).

 -- Benjamin Drung <bdrung@ubuntu.com>  Wed, 14 Oct 2009 20:29:18 +0200

abraca (0.4-2) unstable; urgency=low

  * Add quilt as patch system and state this in debian/README.source.
  * Add debian/patches/version-fix.patch to remove the WIP from the upstream
    version string.
  * Add debian/patches/fix-binaryblob-comment.patch to hopefully fix the
    FTBFS on sparc (Closes: #543251).
  * Allow uploads done by Debian Maintainers.

 -- Benjamin Drung <bdrung@ubuntu.com>  Tue, 25 Aug 2009 18:04:52 +0200

abraca (0.4-1) unstable; urgency=medium

  * New upstream release (Closes: #538380)
    + Compatible to xmms2 0.6DrMattDestruction (Closes: #533988, LP: #394938)
    + Provides an icon (LP: #180014)
  * New maintainer (Closes: #540568)
  * Remove debian/abraca.desktop and debian/abraca.menu. Upstream provides a
    desktop file now.
  * Rewrite debian/rules using debhelper 7.
  * Redo Build-Depends list.
  * Bump Standards-Version to 3.8.3.
  * Add Homepage field.
  * Change priority to optional.
  * Recommend only xmms2 and remove all other recommendations and suggestions.
  * Add debian/watch file.
  * Rewrite debian/copyright from scratch using DEP 5 proposal.
  * Remove debian/abraca.xml, upstream provides a man page.

 -- Benjamin Drung <bdrung@ubuntu.com>  Sat, 22 Aug 2009 22:04:29 +0200

abraca (0.2-2) unstable; urgency=low

  * Recommend essential xmms2 packages and suggest the CLI client.
    (Closes: #456369)

 -- Thomas Jollans <thomas@jollans.com>  Sat, 29 Dec 2007 11:12:17 +0100

abraca (0.2-1.1) unstable; urgency=low

  * Non-maintainer upload.
  * Use a temporary $HOME during build to stop waf writing outside the build
    directory (Closes: #439567).

 -- Simon McVittie <smcv@ianadd.pseudorandom.co.uk>  Mon, 10 Dec 2007 19:47:31 +0000

abraca (0.2-1) unstable; urgency=low

  * Initial release (Closes: #430729)

 -- Thomas Jollans <thomas@jollans.com>  Thu, 19 Jul 2007 22:58:21 +0200



src/abraca.c
/*
 * Copyright (c) 2007 Sebastian Sareyko <smoon at nooms dot de>
 * See COPYING file for details.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <xmmsclient/xmmsclient-glib.h>
#include <glade/glade.h>
#include <gtk/gtk.h>
#include <stdlib.h>
#include <unistd.h>

#include "abraca.h"
#include "collections.h"
#include "filter.h"
#include "medialib.h"
#include "misc.h"
#include "playlist.h"

/* YAY for global variables! */
GladeXML *glade_xml = NULL;
xmmsc_connection_t *xcon = NULL;
track current_track;
gchar *active_playlist = NULL;
guint playlist_position = 0;

/* FIXME: creating some parent containers and list them here would be
 * nicer, but unfortunately Glade looses signal handler names and other
 * stuff on copy/pasting. bleh */
static const gchar *widgets[] = {
    "toolbar_hbox",
    /*
    "track_label",
    "cover_image",
    "time_label",
    "position_scale",
    "fwd_button",
    "back_button",
    "stop_button",
    */
    "playlist_treeview",
    "collections_treeview",
    "filter_treeview",
    "filter_entry",
    "playlist_label",
    "filter_label",
    "playlist_menu_item",
    "add_menu_item",
    "filter_save_menu_item",
    NULL
};

static gboolean xmmscom_poll_playback_playtime (gpointer data);
static xmmsc_connection_t *xmmscom_connect (void);
static void disconnect_callback (void *user_data);
static void setup_callbacks (void);


static gboolean
xmmscom_poll_playback_playtime (gpointer data)
{
    if (xcon) {
        XMMS_CALLBACK_SET (xcon, xmmsc_signal_playback_playtime,
                cb_playback_playtime, NULL);
    }
    else {
        xcon = xmmscom_connect ();
        setup_callbacks ();
    }

    return TRUE;
}

static xmmsc_connection_t
*xmmscom_connect (void)
{
    xmmsc_connection_t *con;
    gint i;

    con = xmmsc_init ("abraca");
    if (!con) {
        for (i = 0; widgets[i]; i++)
            widget_set_sensitive_by_name (widgets[i], FALSE);

        return NULL;
    }

    if (!xmmsc_connect (con, getenv ("XMMS_PATH"))) {
        for (i = 0; widgets[i]; i++)
            widget_set_sensitive_by_name (widgets[i], FALSE);

        con = NULL;
    }
    else {
        for (i = 0; widgets[i]; i++)
            widget_set_sensitive_by_name (widgets[i], TRUE);

        xmmsc_mainloop_gmain_init (con);
    }

    return con;
}

static void
disconnect_callback (void *user_data)
{
    gint i;

    xcon = NULL;

    for (i = 0; widgets[i]; i++)
        widget_set_sensitive_by_name (widgets[i], FALSE);

    playlist_model_clear ();
    filter_model_clear ();
    collections_model_clear ();
}

static void
setup_callbacks (void)
{
    xmmsc_result_t *r;

    if (xcon) {
        XMMS_CALLBACK_SET (xcon, xmmsc_broadcast_playback_current_id,
                bc_playback_current_id, xcon);
        XMMS_CALLBACK_SET (xcon, xmmsc_broadcast_playback_status,
                bc_playback_status, xcon);
        XMMS_CALLBACK_SET (xcon, xmmsc_broadcast_playlist_current_pos,
                bc_playlist_current_pos, xcon);
        XMMS_CALLBACK_SET (xcon, xmmsc_broadcast_playlist_changed,
                bc_playlist_changed, xcon);
        XMMS_CALLBACK_SET (xcon, xmmsc_broadcast_playlist_loaded,
                bc_playlist_loaded, xcon);
        XMMS_CALLBACK_SET (xcon, xmmsc_broadcast_collection_changed,
                bc_collection_changed, xcon);

        xmmsc_disconnect_callback_set (xcon, disconnect_callback, NULL);

        XMMS_CALLBACK_SET (xcon, xmmsc_playback_current_id,
                bc_playback_current_id, xcon);
        r = xmmsc_coll_list (xcon, XMMS_COLLECTION_NS_COLLECTIONS);
        xmmsc_result_notifier_set (r, cb_coll_list,
                XMMS_COLLECTION_NS_COLLECTIONS);
        xmmsc_result_unref (r);
        r = xmmsc_coll_list (xcon, XMMS_COLLECTION_NS_PLAYLISTS);
        xmmsc_result_notifier_set (r, cb_coll_list,
                XMMS_COLLECTION_NS_PLAYLISTS);
        xmmsc_result_unref (r);
        XMMS_CALLBACK_SET (xcon, xmmsc_playlist_current_active,
                cb_playlist_current_active, xcon);
    }
}

gint
main (gint argc, gchar **argv)
{
    gchar *tmp, cwd[1024];
    GtkWidget *mainwin;

#ifdef ENABLE_NLS
    setlocale (LC_ALL, "");
    bindtextdomain ("abraca", LOCALE_DIR);
    textdomain ("abraca");
#endif

    track_nullify (&current_track);

    gtk_init (&argc, &argv);

    /* Glade/GTK+ init stuff */
    tmp = g_build_path (PATHSEP, PACKAGE_DATA_DIR, "abraca.glade", NULL);
    glade_xml = glade_xml_new (tmp, NULL, NULL);
    g_free (tmp);
    /* yay, glade filesearching schmoo */
    if (!glade_xml) {
        getcwd (cwd, sizeof (cwd));
        tmp = g_build_path (PATHSEP, cwd, "data", "abraca.glade",
                NULL);
        glade_xml = glade_xml_new (tmp, NULL, NULL);
        g_free (tmp);

        if (!glade_xml) {
            g_error (_("Failed to load glade file!\n"));
            return (EXIT_FAILURE);
        }
    }

    mainwin = glade_xml_get_widget (glade_xml, "main_window");
    gtk_widget_show (mainwin);

    playlist_view_init ();
    filter_view_init ();
    collections_view_init ();

    glade_xml_signal_autoconnect (glade_xml);

    /* initialize xmms2 */
    xcon = xmmscom_connect ();

    /* xmms2 callbacks */
    setup_callbacks ();

    /* FIXME: use this instead with glib 2.14
       g_timeout_add_seconds (1, poll_playback_playtime, xcon);
    */
    g_timeout_add (1000, xmmscom_poll_playback_playtime, xcon);

    gtk_main ();

    if (xcon)
        xmmsc_unref (xcon);

    return (EXIT_SUCCESS);
}
